test/ecto/query/builder/dynamic_test.exs

```diff
@@ -66,5 +72,29 @@ defmodule Ecto.Query.Builder.DynamicTest do
              "&0.foo() == ^0 and &1.bar() == ^1 and &0.baz() == ^2"
       assert params == [{"foo", {0, :foo}}, {"bar", {1, :bar}}, {"baz", {0, :baz}}]
     end
+
+    test "with ... binding in the middle" do
```

Comment:
Those tests are beautiful! I would recommend moving them to `query_test.exs` as there is nothing specific to fragments about them. Have them work directly on the query and then check the result of `query.joins` (or `query.aliases` according to my proposal).